payments-ui-core/src/main/java/com/stripe/android/ui/core/FormUI.kt

```diff
@@ -74,82 +75,161 @@ fun FormUI(
+                top = 8.dp,
+                bottom = 4.dp,
+            ),
         )
         is AuBecsDebitMandateTextElement -> AuBecsDebitMandateElementUI(
             element = element,
-            modifier = Modifier.padding(vertical = 8.dp),
+            modifier = Modifier.addFormVerticalPadding(
+                size = size,
+                index = index,
+                vertical = 8.dp,
+            ),
         )
         is AffirmHeaderElement -> AffirmElementUI(
-            modifier = Modifier.padding(vertical = 8.dp),
+            modifier = Modifier.addFormVerticalPadding(
+                size = size,
+                index = index,
+                vertical = 8.dp,
+            ),
         )
         is MandateTextElement -> MandateTextUI(
             element = element,
-            modifier = Modifier.padding(top = element.topPadding, bottom = 8.dp)
+            modifier = Modifier.addFormVerticalPadding(
+                size = size,
+                index = index,
+                top = element.topPadding,
+                bottom = 8.dp,
+            ),
         )
         is CardDetailsSectionElement -> CardDetailsSectionElementUI(
-            enabled,
-            element.controller,
-            hiddenIdentifiers,
-            lastTextFieldIdentifier
+            enabled = enabled,
+            controller = element.controller,
+            hiddenIdentifiers = hiddenIdentifiers,
+            lastTextFieldIdentifier = lastTextFieldIdentifier,
+            modifier = Modifier.addFormVerticalPadding(
+                size = size,
+                index = index,
+                vertical = 8.dp,
+            ),
         )
         is BsbElement -> BsbElementUI(
             enabled = enabled,
             element = element,
             lastTextFieldIdentifier = lastTextFieldIdentifier,
-            modifier = Modifier.padding(vertical = 8.dp),
+            modifier = Modifier.addFormVerticalPadding(
+                size = size,
+                index = index,
+                vertical = 8.dp,
+            ),
         )
         is OTPElement -> OTPElementUI(enabled, element)
         is RenderableFormElement -> element.ComposeUI(enabled)
-        is EmptyFormElement -> {}
     }
 }
+
+private fun Modifier.addFormVerticalPadding(
+    index: Int,
+    size: Int,
+    vertical: Dp,
+) = addFormVerticalPadding(
+    index = index,
+    size = size,
+    top = vertical,
+    bottom = vertical
+)
+
+private fun Modifier.addFormVerticalPadding(
+    index: Int,
+    size: Int,
+    top: Dp,
+    bottom: Dp,
+    start: Dp = 0.dp,
+    end: Dp = 0.dp,
+) = when {
+    index == 0 && size == 1 -> this.padding(top = 0.dp, bottom = 0.dp, start = start, end = end)
+    index == 0 -> this.padding(top = 0.dp, bottom = bottom, start = start, end = end)
+    index == size - 1 -> this.padding(top = top, bottom = 0.dp, start = start, end = end)
+    else -> this.padding(top = top, bottom = bottom, start = start, end = end)
+}
```

Comment:
To better explain this logic:
- if only one element, don't apply any top or bottom padding
- else if first element, apply only bottom padding
- else if last element, apply only top padding
- else apply both top and bottom padding